### PR TITLE
Adding controls(reorder) to histogram plot

### PR DESCRIPTION
**Pull request by Fidel365**
fixes issue #8335 ,
@EstherNjeriLiberatta, you can test it now

### CODE REVIEW

**Comment by rdstern** [comment]
image

I tried with the above:
a) The reverse control is not needed on the dialog.  It is a special case of the choices below.  
b) When using the choices  for example as shown - they currently make no difference to the code.  The graph remains the same.  I show the code below:

```
# Dialog: Histogram Plot

survey <- data_book$get_data_frame(data_name="survey")
last_graph <- ggplot2::ggplot(data=survey, mapping=ggplot2::aes(x=yield, fill=variety)) + ggplot2::geom_histogram(mapping=ggplot2::aes(y=stat(count))) + theme_grey()
data_book$add_object(data_name="survey", object_name="last_graph", object_type_label="graph", object_format="image", object=check_graph(graph_object=last_graph))
data_book$get_object_data(data_name="survey", object_name="last_graph", as_file=TRUE)
rm(list=c("last_graph", "survey"))
```

Whatever option I use here, the code remains the same.  Just as for the bar charts, the code for the factor variable should change to reflect the option in the dialog.

(Now the bar chart is a diffficult dialog.  This should be easy by comparison!)

**Comment by Fidel365** [comment]
@rdstern , you can test it now!

**Review by rdstern — CHANGES_REQUESTED** [review]
@Fidel365 Here is the old 0.7.6 with just carat from the diamonds data selected:

image

I am keen that we improve from this starting point.  So here is your new version.  You may be asking me to check too soon, because I would like you, as the developer to do the initial obvious checks.

image

I hope you see that there are already 2 obvious problems - steps backward too - in your improved version:

a) Ok is now not enabled.  It should be.  The factor is a luxury, and you can ask for a single histogram.
b) What is that extra control doing visible there?

So can you enable Ok, when the factor is empty, and only add the properly labelled new Check box when the factor is completed.

Then - it does now seem to work, so well done.  That's a relief.
Can you make it consistent with the bar chart options.  So add the reverse option and ensure the options are in the same order in each dialog.  I am surprised you are not just calling the same function?

**Comment by Fidel365** [comment]
@rdstern , this is ready for a rereview

**Review by rdstern — DISMISSED** [review]
@Fidel365 well done.  That looks ok now.  @lloyddewit over to you.

File: instat/dlgHistogram.vb [hunk 1]
```diff
@@ -168,6 +193,27 @@ Public Class dlgHistogram
         clsPercentage.SetPackageName("scales")
         clsPercentage.SetRCommand("percent_format")
 
+        clsGeomTextFunction.SetPackageName("ggplot2")
+        clsGeomTextFunction.SetRCommand("geom_text")
+        'clsGeomTextFunction.AddParameter("mapping", clsRFunctionParameter:=clsLabelAesFunction, iPosition:=1)
```
Comment (lloyddewit) [inline instat/dlgHistogram.vb:198]:
```suggestion
```

File: instat/dlgHistogram.vb [hunk 2]
```diff
@@ -200,14 +246,15 @@ Public Class dlgHistogram
...
         End If
     End Sub
 
     Private Sub TestOkEnabled()
         'Tests when ok can be enabled
-        If ucrVariablesAsFactorforHist.IsEmpty OrElse Not ucrSaveHist.IsComplete OrElse (ucrChkRidges.Checked AndAlso (ucrFactorReceiver.IsEmpty OrElse ucrFactorReceiver.Enabled = False)) Then
+        If ucrVariablesAsFactorforHist.IsEmpty OrElse Not ucrSaveHist.IsComplete Then
             ucrBase.OKEnabled(False)
         Else
             ucrBase.OKEnabled(True)
```
Comment (lloyddewit) [inline instat/dlgHistogram.vb:260]: Is this simpler?
```suggestion
        ucrBase.OKEnabled(Not ucrVariablesAsFactorforHist.IsEmpty AndAlso ucrSaveHist.IsComplete)
```

File: instat/dlgHistogram.vb [hunk 3]
```diff
@@ -225,51 +272,72 @@ Public Class dlgHistogram
...
             ucrFactorReceiver.ChangeParameterName("fill")
-            If Not ucrSaveHist.bUserTyped Then ucrSaveHist.SetPrefix("histogram")
+            '    If ucrChkReverse.Checked Then
+            '        clsForecatsReverseValue.AddParameter("f", ucrFactorReceiver.GetVariableNames(False), iPosition:=0)
+            '        clsRaesFunction.AddParameter("fill", clsRFunctionParameter:=clsForecatsReverseValue, iPosition:=1)
+            '    Else
+            '        clsRaesFunction.AddParameter("fill", ucrFactorReceiver.GetVariableNames(False), iPosition:=1)
+            '    End If
+            '    If Not ucrSaveHist.bUserTyped Then ucrSaveHist.SetPrefix("histogram")
+            '    ucrChkReverse.Visible = Not ucrFactorReceiver.IsEmpty()
```
Comment (lloyddewit) [inline instat/dlgHistogram.vb:290]:
```suggestion
```

File: instat/dlgHistogram.vb [hunk 4]
```diff
@@ -225,51 +272,72 @@ Public Class dlgHistogram
...
         End If
         If rdoDensity_ridges.Checked Then
+            ucrFactorReceiver.ChangeParameterName("fill")
+            'If ucrChkReverse.Checked Then
+            '    clsForecatsReverseValue.AddParameter("f", ucrFactorReceiver.GetVariableNames(False), iPosition:=0)
+            '    clsRaesFunction.AddParameter("fill", clsRFunctionParameter:=clsForecatsReverseValue, iPosition:=1)
+            'Else
+            '    clsRaesFunction.AddParameter("fill", ucrFactorReceiver.GetVariableNames(False), iPosition:=1)
+            'End If
+            'ucrChkReverse.Visible = Not ucrFactorReceiver.IsEmpty()
```
Comment (lloyddewit) [inline instat/dlgHistogram.vb:300]:
```suggestion
```

File: instat/dlgHistogram.vb [hunk 5]
```diff
@@ -225,51 +272,72 @@ Public Class dlgHistogram
...
+                ucrSaveHist.SetPrefix("density")
+            End If
+            ucrFactorReceiver.ChangeParameterName("fill")
+            'If ucrChkReverse.Checked Then
+            '    clsForecatsReverseValue.AddParameter("f", ucrFactorReceiver.GetVariableNames(False), iPosition:=0)
+            '    clsRaesFunction.AddParameter("fill", clsRFunctionParameter:=clsForecatsReverseValue, iPosition:=1)
+            'Else
+            '    clsRaesFunction.AddParameter("fill", ucrFactorReceiver.GetVariableNames(False), iPosition:=1)
+            'End If
+            'ucrChkReverse.Visible = Not ucrFactorReceiver.IsEmpty()
```
Comment (lloyddewit) [inline instat/dlgHistogram.vb:328]:
```suggestion
```

File: instat/dlgHistogram.vb [hunk 6]
```diff
@@ -382,4 +458,30 @@ Public Class dlgHistogram
...
+                    clsHistAesFunction.AddParameter("fill", clsRFunctionParameter:=clsForecatsReverse, iPosition:=1)
+                Case strDescending
+                    clsHistAesFunction.AddParameter("fill", clsRFunctionParameter:=clsForecatsInfreqValue, iPosition:=1)
+                Case strReverse
+                    clsForecatsReverse.AddParameter("f", ucrFactorReceiver.GetVariableNames(False), iPosition:=0)
+                    clsHistAesFunction.AddParameter("fill", clsRFunctionParameter:=clsForecatsReverse, iPosition:=0)
+            End Select
+        End If
+    End Sub
+    Private Sub ucrPnlOptions_Control(ucrChangedControl As ucrCore) Handles ucrPnlOptions.ControlValueChanged, ucrFactorReceiver.ControlValueChanged,
```
Comment (lloyddewit) [inline instat/dlgHistogram.vb:479]:
```suggestion

    Private Sub ucrPnlOptions_Control(ucrChangedControl As ucrCore) Handles ucrPnlOptions.ControlValueChanged, ucrFactorReceiver.ControlValueChanged,
```

File: instat/dlgHistogram.vb [hunk 7]
```diff
@@ -382,4 +458,30 @@ Public Class dlgHistogram
...
+        ucrChkRidges.ControlValueChanged, ucrChkDisplayAsDotPlot.ControlValueChanged, ucrVariablesAsFactorforHist.ControlValueChanged,
+        ucrInputAddReorder.ControlValueChanged
+        UpdateParameter()
+    End Sub
+
+    Private Sub CoreControls_ControlContentsChanged(ucrChangedControl As ucrCore) Handles ucrVariablesAsFactorforHist.ControlContentsChanged, ucrSaveHist.ControlContentsChanged, ucrFactorReceiver.ControlContentsChanged, ucrChkRidges.ControlContentsChanged
+    End Sub
```
Comment (lloyddewit) [inline instat/dlgHistogram.vb:486]: duplicate? delete?
```suggestion
```

**Review by lloyddewit — COMMENTED** [review]
@Fidel365 thanks, just a few small comments

**Comment by Fidel365** [comment]
@lloyddewit , done!

**Review by rdstern — CHANGES_REQUESTED** [review]
@Fidel365 I had hoped we were through, but trying more systematically there are still 2 sorts of problem in your R-code.
a) It is sometimes incorrect - you do need to make sure you understand the R code a bit for ggplot, because it will usually give something and you must make sure that it is the right thing!
b) Also, and this is a VB issue.  Make sure that addition e.g. reverse, are clearing when you take them away again.

We should skype when you are ready.  I assume not over the weekend for you, but then hopefully Monday morning?  

I am going to explain in detail so you are able to make the changes.  And please make sure you are learning and understanding more through this, so you can make changes more quickly and reliably in the future.

Here is my starting point, with the ggplot2 diamonds data:

image

a) And here is the code you generate:

```
# Dialog: Histogram Plot

diamonds <- data_book$get_data_frame(data_name="diamonds")
last_graph <- ggplot2::ggplot(data=diamonds, mapping=ggplot2::aes(x=carat, fill=cut)) + ggplot2::geom_histogram(mapping=ggplot2::aes(y=stat(count))) + theme_grey()
data_book$add_object(data_name="diamonds", object_name="last_graph", object_type_label="graph", object_format="image", object=check_graph(graph_object=last_graph))
data_book$get_object_data(data_name="diamonds", object_name="last_graph", as_file=TRUE)
rm(list=c("last_graph", "diamonds"))
```
b) Now I change to the reverse option:  And the line of code is changes to:

`last_graph <- ggplot2::ggplot(data=diamonds, mapping=ggplot2::aes(x=carat, fill=cut)) + ggplot2::geom_histogram(mapping=ggplot2::aes(fill=forcats::fct_rev(f=cut), y=stat(count))) + theme_grey()`

That's ok, but I wonder why you added the second fill in the histogram geom.  You could have changed it in the ggplot2::aes(x=carat, fill=cut)) bit instead?  The fact you are adding these parameters is messing you up.  Remember if you repeat a parameter, then it will be the last occurrence that counts. 

c) Now I take the reverse off - so I am back to the start.  Your code is now:

`last_graph <- ggplot2::ggplot(data=diamonds, mapping=ggplot2::aes(x=carat, fill=cut)) + ggplot2::geom_histogram(mapping=ggplot2::aes(y=stat(count), fill=forcats::fct_rev(f=cut))) + theme_grey()`

Notice your reverse is still there!

So that's the VB problem.  Make sure that the code you add is subtracted, when it is taken away.
 
Now press reset, because I think the reverse - and perhaps others - stay forever!

Then try ridge plots, first with nothing.  
image

This gives gray plots.  And code:

```
# Dialog: Histogram Plot

diamonds <- data_book$get_data_frame(data_name="diamonds")
carat <- data_book$get_columns_from_data(data_name="diamonds", col_names="carat")
cut <- data_book$get_columns_from_data(data_name="diamonds", col_names="cut")
last_graph <- ggplot2::ggplot(data=diamonds, mapping=ggplot2::aes(x=carat, y=cut)) + ggridges::geom_density_ridges(mapping=ggplot2::aes(x=carat, y=cut)) + theme_grey()
data_book$add_object(data_name="diamonds", object_name="last_graph", object_type_label="graph", object_format="image", object=check_graph(graph_object=last_graph))
data_book$get_object_data(data_name="diamonds", object_name="last_graph", as_file=TRUE)
rm(list=c("last_graph", "diamonds", "carat", "cut"))

```
I don't know whether you repeated the parameters in the ggplot2 line, but

`ggridges::geom_density_ridges()`  is simpler and works just fine!

Now the line changes with reverse to the following:  

`last_graph <- ggplot2::ggplot(data=diamonds, mapping=ggplot2::aes(x=carat, y=cut)) + ggridges::geom_density_ridges(mapping=ggplot2::aes(fill=forcats::fct_rev(f=cut), x=carat, y=cut)) + theme_grey()
`
If you are following the logic, you should see why that's incorrect.  I should also have noticed that the fact it leaps into colour indicated a real problem.

This has been useful though as I now see whay what I thought was a simple change has gone on and on.  Partly my fault, not looking carefully enough before!

**Review by rdstern — CHANGES_REQUESTED** [review]
@Fidel365 I started with histogram, then added a factor, and it was fine.  Then changed to reverse and it seems the same as before.  This is acceptable.  Then I deleted the factor and it is still there in the code.  So the aspect of clearing - at least for this instance is still not solved.  

I then went to the density option.  Look below.  Did you not notice this was happening - there is no factor in the dialog, but it is still there in the code.

image

I then went to the density plots and they seem to be the same as before.  We discussed all this on Sunday, and I don't see what's changed.  Is this still "work in progress" in which case please say so, then I'll wait before testing?

I thought you had agreed with me, on Sunday, that this is a very simple set of changes.  If they are still difficult for you to do, I suggest you ask @N-thony to allocate it to someone else?  If so, then I would like to explain what's needed to them, before they start work.

**Comment by Fidel365** [comment]
@rdstern I'll commit today

**Comment by rdstern** [comment]
@Fidel365 I am not sure what you have changed here.  But the many problems seem to remain.  @N-thony I seek your advice, because, unless I have misunderstood this is taking Fidel a lot of time and it is not going to succeed.  My inclination is to give him other tasks and give Sophie or Vitalis this work.  If you do, then I am unsure whether they should start from this dialog, or simply add the reorder control to the original histogram code.  

I would then like to explain the work and also Fidel should be a checker of the results.  I would also like whoever does this to keep an account of the time.  I suggest - certainly if you were to do it, that it should take perhaps 2 hours, but I may be wrong!  I think Sophie may need 3 hours and Vitalis may need 4 hours.

**Comment by Fidel365** [comment]
@rdstern, I'm actually working on this with Vitalis, I'll push the recent commits as soon as tomorrow before our standup meeting

**Comment by Fidel365** [comment]
PR #8547 fixes the issue

**Pull request closed without merging**